### ISSUE TITLE
تمرین پیدا کردن دیتاست و انجام عمل های دیگر روی آن

### ISSUE THREAD

**Issue by romosij200218-prog** [Amir13849271741537/Machine-learning-#15]
# motor_fault_detection.ipynb
# Python 3.9+ recommended
# Install required packages if needed:
# !pip install numpy pandas matplotlib seaborn scikit-learn scipy joblib tqdm kaggle

# ---------------------------
# 0) Imports & settings
# ---------------------------
import os
import glob
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import signal
from scipy.fft import rfft, rfftfreq
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_auc_score, roc_curve
import joblib
from tqdm import tqdm

# For reproducibility
RND = 42
plt.rcParams["figure.figsize"] = (8,5)

# ---------------------------
# 1) Download / place dataset
# ---------------------------
# Option A: اگر از Kaggle استفاده می‌کنید، در محیط لوکال:
# 1) نصب kaggle و تنظیم api token (kaggle.json) در ~/.kaggle/
# 2) دانلود (مثال برای dataset id):
#    kaggle datasets download -d josh101/machinery-fault-database-induction-motor-fault --unzip -p ./data
#
# Option B: اگر دیتاست رو دارید، فولدر data/ را طوری که فایل‌های csv داخلش باشه قرار بدید.
#
DATA_DIR = "./data"  # مسیر داده‌ها

# ---------------------------
# 2) بارگذاری داده‌ها (چند سناریو)
# ---------------------------
def load_csv_files(folder):
    """Load CSV files in folder into a single DataFrame by concatenation.
       Adjust parsing depending on dataset format."""
    files = glob.glob(os.path.join(folder, "**", "*.csv"), recursive=True)
    if not files:
        return None
    dfs = []
    for f in files:
        try:
            df = pd.read_csv(f)
            df['source_file'] = os.path.basename(f)
            dfs.append(df)
        except Exception as e:
            print("Could not read", f, e)
    if dfs:
        return pd.concat(dfs, ignore_index=True)
    return None

df = load_csv_files(DATA_DIR)
if df is None:
    print("No CSV files found in ./data. Generating a synthetic demo dataset so you can run pipeline.")
    # ---------------------------
    # Synthetic demo dataset (برای تست کد)
    # ---------------------------
    from sklearn.datasets import make_classification
    X_s, y_s = make_classification(n_samples=2000, n_features=60, n_informative=15,
                                   n_redundant=5, n_classes=2, random_state=RND)
    df = pd.DataFrame(X_s, columns=[f"x{i}" for i in range(X_s.shape[1])])
    df['label'] = y_s
    # mark as already feature matrix (skip feature-extraction)
    SYNTHETIC = True
else:
    SYNTHETIC = False
    print("Loaded dataframe shape:", df.shape)
    display(df.head())

# ---------------------------
# 3) اگر دیتاست اصلی زمان-سری است: استخراج ویژگی‌ها
#    (در بسیاری از دیتاست‌های موتور، هر رکورد یک سیگنال کوتاه است)
# ---------------------------
def extract_time_domain_features(signal_array):
    # signal_array: 1D numpy array
    features = {}
    features['mean'] = np.mean(signal_array)
    features['std'] = np.std(signal_array)
    features['var'] = np.var(signal_array)
    features['max'] = np.max(signal_array)
    features['min'] = np.min(signal_array)
    features['median'] = np.median(signal_array)
    features['skew'] = pd.Series(signal_array).skew()
    features['kurtosis'] = pd.Series(signal_array).kurtosis()
    features['rms'] = np.sqrt(np.mean(signal_array**2))
    # peak-to-peak
    features['ptp'] = np.ptp(signal_array)
    # zero-crossing rate
    zc = ((signal_array[:-1] * signal_array[1:]) < 0).sum()
    features['zc_rate'] = zc / len(signal_array)
    return features

def extract_freq_domain_features(signal_array, fs=1.0, n_peaks=5):
    # FFT-based simple features
    N = len(signal_array)
    yf = np.abs(rfft(signal_array))
    xf = rfftfreq(N, 1/fs)
    # dominant frequency and its magnitude
    idx = np.argmax(yf)
    features = {}
    features['dom_freq'] = xf[idx]
    features['dom_mag'] = yf[idx]
    # spectral centroid
    sc = np.sum(xf * yf) / (np.sum(yf) + 1e-9)
    features['spec_centroid'] = sc
    # spectral entropy (simple normalized)
    ps = yf / (np.sum(yf) + 1e-9)
    spec_entropy = -np.sum(ps * np.log(ps + 1e-12))
    features['spec_entropy'] = spec_entropy
    return features

# If synthetic, we skip; else we need to detect how the real dataset is structured.
if not SYNTHETIC:
    # Heuristic: if dataframe has a 'label' column, and many columns are sample points -> treat as features
    if 'label' in df.columns:
        print("Dataset includes 'label' column -> treating each row as pre-extracted features or whole-signal collapsed.")
        feature_df = df.copy()
    else:
        # Example: if each CSV file is a time series with columns ['time','acc_x','acc_y','acc_z','label'] and many rows,
        # we should segment signals and extract features per segment. Below is a generic placeholder process.
        print("No 'label' column detected. Attempting to assemble features by grouping source_file or segmenting.")
        # This will highly depend on the dataset format — adjust accordingly.
        # For safety, here we just take numeric columns and compute their statistics per source_file.
        numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
        groups = df.groupby('source_file')
        feats = []
        labels = []
        for name, g in groups:
            row_feats = {}
            for col in numeric_cols:
                arr = g[col].values
                ft = extract_time_domain_features(arr)
                for k,v in ft.items():
                    row_feats[f"{col}_{k}"] = v
            # if label present in group use mode
            if 'label' in g.columns:
                labels.append(g['label'].mode()[0])
            else:
                labels.append(0)  # placeholder
            feats.append(row_feats)
        feature_df = pd.DataFrame(feats)
        feature_df['label'] = labels

# For synthetic we already have dataframe with features
if SYNTHETIC:
    feature_df = df.copy()

print("Feature matrix shape:", feature_df.shape)
display(feature_df.head())

# ---------------------------
# 4) آماده‌سازی X, y
# ---------------------------
# assume last column 'label' holds target (adjust name if different)
target_col = 'label'
if target_col not in feature_df.columns:
    raise ValueError("Target column 'label' not found. Adjust target_col variable.")

X = feature_df.drop(columns=[target_col]).select_dtypes(include=[np.number]).fillna(0)
y = feature_df[target_col].values

# encode label if not numeric
if X.shape[0] > 0 and (y.dtype == object or y.dtype == '<U'):
    le = LabelEncoder()
    y = le.fit_transform(y)

print("X shape:", X.shape, "y shape:", y.shape)
print("Classes:", np.unique(y))

# ---------------------------
# 5) تقسیم داده و مقیاس‌دهی
# ---------------------------
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=RND)
sc = StandardScaler()
X_train_s = sc.fit_transform(X_train)
X_test_s = sc.transform(X_test)

# ---------------------------
# 6) مدل‌ها: آموزش و ارزیابی
# ---------------------------
models = {
    "RandomForest": RandomForestClassifier(n_estimators=200, random_state=RND),
    "SVM": SVC(kernel='rbf', probability=True, random_state=RND),
    "MLP": MLPClassifier(hidden_layer_sizes=(100,50), max_iter=300, random_state=RND)
}

results = {}
for name, model in models.items():
    print(f"\nTraining {name} ...")
    if name == "RandomForest":
        model.fit(X_train, y_train)  # tree-based doesn't need scaling
    else:
        model.fit(X_train_s, y_train)
    # predict on test
    if name == "RandomForest":
        y_pred = model.predict(X_test)
        if hasattr(model, "predict_proba"):
            y_proba = model.predict_proba(X_test)[:,1] if len(np.unique(y))==2 else None
    else:
        y_pred = model.predict(X_test_s)
        y_proba = model.predict_proba(X_test_s)[:,1] if len(np.unique(y))==2 else None

    acc = accuracy_score(y_test, y_pred)
    print(f"{name} Test Accuracy: {acc:.4f}")
    print(classification_report(y_test, y_pred, zero_division=0))
    cm = confusion_matrix(y_test, y_pred)
    print("Confusion matrix:\n", cm)
    results[name] = {"model": model, "acc": acc, "y_pred": y_pred, "y_proba": y_proba}

# ---------------------------
# 7) مقایسه مدل‌ها (نمودار دقت)
# ---------------------------
names = list(results.keys())
accs = [results[n]['acc'] for n in names]
plt.figure(figsize=(6,4))
plt.bar(names, accs)
plt.title("Model Accuracy Comparison")
plt.ylabel("Accuracy")
for i,v in enumerate(accs):
    plt.text(i, v+0.01, f"{v:.3f}", ha='center')
plt.ylim(0,1.05)
plt.show()

# ---------------------------
# 8) ROC (if binary)
# ---------------------------
if len(np.unique(y)) == 2:
    plt.figure()
    for name in names:
        y_proba = results[name]['y_proba']
        if y_proba is None:
            continue
        fpr, tpr, _ = roc_curve(y_test, y_proba)
        auc = roc_auc_score(y_test, y_proba)
        plt.plot(fpr, tpr, label=f"{name} (AUC={auc:.3f})")
    plt.plot([0,1],[0,1],'--')
    plt.xlabel("FPR")
    plt.ylabel("TPR")
    plt.legend()
    plt.title("ROC Curves")
    plt.show()

# ---------------------------
# 9) Cross-validation on best model (مثال: RandomForest)
# ---------------------------
best = max(results.items(), key=lambda x: x[1]['acc'])[0]
print("Best model:", best)
clf = models[best]
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RND)
if best == "RandomForest":
    scores = cross_val_score(clf, X, y, cv=cv, scoring='accuracy')
else:
    # pipeline for scaling
    from sklearn.pipeline import make_pipeline
    pipeline = make_pipeline(StandardScaler(), clf)
    scores = cross_val_score(pipeline, X, y, cv=cv, scoring='accuracy')
print("5-fold CV accuracies:", scores)
print("CV mean:", scores.mean(), "std:", scores.std())

# ---------------------------
# 10) ذخیره مدل و scaler
# ---------------------------
joblib.dump(results[best]['model'], "best_model.joblib")
joblib.dump(sc, "scaler.joblib")
print("Saved best model and scaler.")
